docs/arch/integrations/adapters/evm/interop/motivation.md

```diff
@@ -0,0 +1,81 @@
+
+That means, that what we want is not just to make a call, but to make a call subject to a certain logic. Moreover, we may want to bind it not just to a logic, but to the label, as kinds play an important role in application design.
+
+Therefore we need not only the original contract, but a place to store both logic and label connected to the application. For that, we need to have a Forwarder Contract. It will be a contract that can perform the needed calls with arbitrary constraints and store the needed resource kind making sure only specific application will be able to perform possible operations through them.
+
+A Forwarder Contract is a contract which has two functions: `calldataCarrierResourceKind` and `forwardCall`
+
+`forwardCall` will be called by the PA, well, in order to perform arbitrary logic with inputs specified.
+
+`calldataResourceKind` is a `view` function that exposes to the RM the kind of the resource that can specify a call to be made to the Forwarder.
+
+### Calldata Resources
+
+Interoperability requires actors on both sides: on the RM and EVM. On the EVM we need the Forwarder Contracts to allow arbitrary calls to be made by application developers bound to a specific application. On the RM side, we need resources representing an application to verify that a call is being made.
+
+For that, we need to store the information regarding a call inside the context of the resource logics being evaluated. In partciular, we have the `externalPayload` field inside `appData`. There we put information such as the address of the forwarder contract we call, inputs to the call, as well as the expected outputs.
+
+As `appData` gets exposed to the resource logics they can make sure that a call is present (or absent) and will be made by the PA, returning the desired output.
+
+At the same time the PA makes sure that the call is accompanied by the appropriate resource using the `calldataResourceKind` function of the contract. For that `appData` also is expetced to conatain some resource plaintext information. That way we will be able to check that the plaintext has the kind specified by the corresponding forwarder contract.
+
+Here is approximately the logic of the PA to locate these calls and make them after running `verify` on the transaction, ensuring in particular that all RM-specified checks pass:
+
+Go through each `appData.externalPayload` of each `tag`. Recover the plantext from `appData.resourcePayload`, and check that `hash(resourcePlaintext) = tag`. This may require the PA also knowing the nullifier key if the resource is being consumed. Finally, we check that `kind(resourcePlaintext)` is the same as the return of `calldataCarrierResourcekind()`.
```

Comment:
```suggestion
Go through each `appData.externalPayload` of each `tag`. Recover the plantext from `appData.resourcePayload`, and check that `hash(resourcePlaintext) = tag`. This may require the PA also knowing the nullifier key if the resource is being consumed. Finally, we check that `kind(resourcePlaintext)` is the same as the return of `calldataCarrierResourceKind()`.
```